### ISSUE TITLE
Is there a way to define nested mapping with interfaces ?

### ISSUE THREAD

**Issue by blemasle** [agileobjects/AgileMapper#172]
Hi,

It's me again with more interfaces issues. Is there a way to map an object to a nested property when using interfaces ? Consider the following use case.

https://dotnetfiddle.net/cwTE25
```csharp
using AgileObjects.AgileMapper;
using System;
using System.Collections.Generic;

namespace MapperSandbox
{
	class Program
	{

		static void Main(string[] args)
		{
			//working 
			Mapper
				.WhenMapping
				.From<A>()
				.To<BWithoutInterface>()
				.Map((ctx) => ctx.Source).To(dst => dst.Nested);

			Mapper
				.WhenMapping
				.From<A>()
				.To<NestedB>()
				.Map((src, dst) => src.Something).To(dst => dst.AnotherProperty);

			BWithoutInterface b = Mapper.Map(new A { Something = "Hello World!" }).ToANew<BWithoutInterface>();
			Console.WriteLine("AnotherProperty: " + b.Nested.AnotherProperty);

			//not working
			Mapper
				.WhenMapping
				.From<A>()
				.To<B>()
				.Map((ctx) => ctx.Source).To(dst => dst.Nested);

			Mapper
				.WhenMapping
				.From<A>()
				.To<INestedB>()
				.CreateInstancesOf<NestedB>();

			//defined above
			//Mapper
			//	.WhenMapping
			//	.From<A>()
			//	.To<NestedB>()
			//	.Map((src, dst) => src.Something).To(dst => dst.AnotherProperty);

			B b2 = Mapper.Map(new A { Something = "Hello World!" }).ToANew<B>();
			Console.WriteLine("AnotherProperty: " + b2.Nested.AnotherProperty);
		}
	}

	class A
	{
		public string Something { get; set; }
	}

	class B
	{
		public INestedB Nested { get; set; }
	}

	class BWithoutInterface
	{
		public NestedB Nested { get; set; }
	}

	public interface INestedB
	{
		string AnotherProperty { get; set; }
	}

	class NestedB : INestedB
	{
		public string AnotherProperty { get; set; }
	}
}
```

I cannot understand the following points :
* `CreateInstancesOf` does not create an instance of the specified type.
* I could use `CreateInstancesUsing<NestedB>()` but in this case, the mapping is not applied either
* I cannot find a way to define the mapping for `B.Nested` without relying on a standard method call (instead of using mapping expressions)

Is there a way to implement my use case ?
Cheers

**Comment by SteveWilkes** [comment]
Hello again!

You can do it as shown in [this fiddle](https://dotnetfiddle.net/Cx1Mvj) - you do need to use `CreateInstancesUsing` instead of `CreateInstancesOf` - the latter is intended to be followed with another call:  `CreateInstancesOf<MyType>().Using(ctx => new MyType())`. I think I need to document that better or error if it doesn't have the `Using` call chained onto it.

Cheers!

Steve

**Comment by blemasle** [comment]
Ohhh I see. I had no trouble to see that `CreateInstancesOf().Using()` and `CreateInstancesUsing()` were indeed the same thing. I understood the API as 

> When mapping from `A` to `INestedB`, "redirect" the mapping to `NestedB` instead
> When mapping from `A` to `NestedB`, map ...

So I defined mappings in two separated instructions. Your use case/fiddle for instance cannot be used with a read only property in the interface and a read/write property in the implementation (which is a very common thing to do).

Another thing that troubles me is why do I need to supply a constructor ? I mean, this is the first time I had to do so. It's nice to have the opportunity to use one, but shouldn't `CreateInstancesOf<NestedB>()` use the same logic as `From<A>().To<NestedB>()` ?

Thanks again,
Bertrand

**Comment by SteveWilkes** [comment]
Hmm - I think I might be missing a useful configuration option here, I'm going to look into it :)

**Comment by SteveWilkes** [comment]
Ok, I think we should be able to address this with `MapTo<T>()`. At the moment, you can do:

```csharp
Mapper.WhenMapping
    .From<A>()
    .To<B>()
    .Map<ADerived>()
    .To<BDerived>()
```

...and:

```csharp
Mapper.WhenMapping
    .From<A>()
    .To<B>()
    .If(ctx => condition)
    .MapTo<BDerived>()
```

The latter case doesn't work as I'd expect if `B` is an interface, so I'll look into that, and I'm also going to support:

```csharp
Mapper.WhenMapping
    .From<A>()
    .To<B>()
    .MapTo<BDerived>()
```

Mapping configuration continuations (`.And`) from `CreateInstancesUsing` and `MapTo` continue configuration of `A` -> `B` mapping at the moment; I'm going to update it to configure `A` -> `BDerived` mapping instead. The current set up is why the readonly interface member configuration doesn't work - we're configuring a source for that readonly member - not the writeable implementation member - so the error is actually appropriate.

I'll get back to you! :)

Steve

**Comment by blemasle** [comment]
Yep, I figured it out by looking at mapping continuations signatures, and that's why I proposed the `Use` signature. I even tried to implement it myself but stumble upon an obscure expression conversion error 😄 
Of course, the `MapTo` you're suggesting is way more in line with the library API 😃 

In the meantime, I changed the interface to accept setter (although not ideal, it works for now) as in my use case I don't have a `ADerived` so the first proposition is not applicable. 

Thanks again !
Bertrand

**Comment by SteveWilkes** [comment]
The use case as described is now supported in the v1.6 release branch! Apologies for the delay - release to follow!

**Comment by SteveWilkes** [comment]
The fix for this is included in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.6.0).

Thanks again for the feedback!
